src/pages/SingleCell/PieToolTip.tsx

```diff
@@ -0,0 +1,393 @@
+            if (pieChartData[j].patientId === id) {
+                patientData[id].push({
+                    stableId: pieChartData[j].stableId,
+                    value: pieChartData[j].value,
+                });
+            }
+        }
+    }
+    useEffect(() => {
+        if (downloadSvg) {
+            handleDownloadSVG(svgRef);
+            setDownloadSvg(false);
+        }
+    }, [downloadSvg]);
+
+    useEffect(() => {
+        if (downloadPdf) {
+            handleDownloadPDF(svgRef);
+            setDownloadPdf(false);
+        }
+    }, [downloadPdf]);
+
+    // Set tooltip visibility with a delay when hover state changes, only if tooltipEnabled is false
+    useEffect(() => {
+        if (tooltipEnabled) {
+            setTooltipVisible(true);
+        } else {
+            if (isHovered || tooltipHovered) {
+                setTooltipVisible(true);
+            } else {
+                const timeoutId = setTimeout(
+                    () => setTooltipVisible(false),
+                    300
+                ); // 300ms delay before hiding tooltip
+                return () => clearTimeout(timeoutId);
+            }
+        }
+    }, [isHovered, tooltipHovered, tooltipEnabled]);
+
+    // Define color scale (replace with your desired colors)
+    const colors = [
+        '#00BCD4', // Cyan (High contrast, good accessibility)
+        '#FF9800', // Orange (Warm, contrasting)
+        '#A52A2A', // Maroon (Deep, high contrast)
+        '#795548', // Brown (Earth tone, contrasts well with previous)
+        '#27AE60', // Pink (Light, good contrast)
+        '#E53935', // Green (Vibrant, contrasts with Pink)
+        '#9C27B0', // Violet (Rich, unique hue)
+        '#2986E2', // Blue (Calming, high contrast)
+        '#FFEB3B', // Light Yellow (Light, good contrast with Blue)
+        '#051288', // Red (Bold, contrasts well)
+        '#008080',
+        '#7a8376',
+    ];
+
+    // Filter out data bins with specialValue "NA"
+    const uniqueIds: any = [
+        ...new Set(pieChartData.map((item: any) => item.genericAssayStableId)),
+    ];
+    const sumValues: { [key: string]: number } = {};
+
+    // Calculate the total sum of sumValues
+    let totalSum = 0;
+    uniqueIds.forEach((id: string) => {
+        sumValues[id] = pieChartData.reduce((acc: number, item: any) => {
+            if (item.genericAssayStableId === id) {
+                const value = parseFloat(item.value);
+                totalSum += value;
+                return acc + value;
+            }
+            return acc;
+        }, 0);
+    });
+
+    const pieData = Object.keys(sumValues).map((key, index) => {
+        const color =
+            pieChartData.find((item: any) => item.genericAssayStableId === key)
+                ?.color || colors[index];
+        return {
+            typeOfCell: key,
+            percentage: sumValues[key],
+            color: color,
+        };
+    });
+
+    const handleDownloadData = () => {

```

Comment:
Why do we need a `handleDownloadData` in the tooltip? Is it possible to only download the tooltip? And like mentioned before, try to generalize this `handleDownloadData`, place it in `downloadUtils` and import it.

Reply:
And please explain why we cannot use the existing download function of cBioPortal

Reply:
I have generalized all the download functions into downloadUtils. The reason I haven't reused the existing download functions is that we need additional customization. For instance, in the histogram and pie chart, we want the downloaded chart to include percentages next to the legends, even though these percentages are not displayed in the chart. Because of these customizations, I wrote specific functions.